akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -137,6 +139,200 @@ import akka.util.ccompat.JavaConverters._
+      } else throw new NoSuchElementException("next called after completion")
+    override def tryPull(): Unit = ()
+    override def cancel(): Unit = ()
+    override def isClosed: Boolean = !hasNext
+  }
+
+  final class InflightIterableSource[T](elements: Iterable[T]) extends InflightSource[T] {
+    private val iterator: Iterator[T] = elements.iterator
+    override def materialize(): Unit =
+      new UnsupportedOperationException("Do not need materialization for InflightSingleSource.")
+    override def hasNext: Boolean = iterator.hasNext
+    override def next: T = iterator.next()
+    override def tryPull(): Unit = ()
+    override def cancel(): Unit = ()
+    override def isClosed: Boolean = !hasNext
+  }
+
+  object InflightEmptySource extends InflightSource[Any] {
+    override def materialize(): Unit =
+      new UnsupportedOperationException("Do not need materialization for InflightSingleSource.")
+    override def hasNext: Boolean = false
+    override def next: Any = throw new NoSuchElementException("next called after completion")
+    override def tryPull(): Unit = ()
+    override def cancel(): Unit = ()
+    override def isClosed: Boolean = true
+  }
+}
+
+@InternalApi
+private[akka] final class FlattenConcat[T, M](parallelism: Int)
+    extends GraphStage[FlowShape[Graph[SourceShape[T], M], T]] {
+  require(parallelism >= 1, "parallelism should >= 1")
+  private val in = Inlet[Graph[SourceShape[T], M]]("flattenConcat.in")
+  private val out = Outlet[T]("flattenConcat.out")
+
+  override def initialAttributes: Attributes = DefaultAttributes.flattenConcat
+  override val shape: FlowShape[Graph[SourceShape[T], M], T] = FlowShape(in, out)
+  override def createLogic(inheritedAttributes: Attributes) =
+    new GraphStageLogic(shape) with InHandler with OutHandler {
+      import FlattenConcat._
+      private var inflightSources: BufferImpl[InflightSource[T]] = _
+
+      override def preStart(): Unit = inflightSources = BufferImpl(parallelism, inheritedAttributes)
+
+      override def onPush(): Unit = {
+        val source = grab(in)
+        addSource(source)
+        if (!inflightSources.isFull) {
+          tryPull(in)
+        }
+      }
+
+      override def onUpstreamFinish(): Unit = if (inflightSources.isEmpty) completeStage()
+
+      override def onPull(): Unit = {
+        pull(in)
+        setHandler(
+          out,
+          new OutHandler {
+            override def onPull(): Unit = {
+              if (inflightSources.nonEmpty) {
+                val currentSource = inflightSources.peek()
+                //purge if possible
+                if (currentSource.hasNext) {
+                  push(out, currentSource.next)
+                  if (currentSource.isClosed) {
+                    removeSource(currentSource)
+                  }
+                } else if (currentSource.isClosed) {
+                  removeSource(currentSource)
+                } else {
+                  currentSource.tryPull()
```

Comment:
As we discussed in private, do you think it is better to move `currentSource.tryPull` into the same branch of `push` and pull `sinkIn` after push element into `out`? since it will more like previous implementation that make it easier to reasoning for members(since it works pretty stable) and delegate the push/pull loop to `subFusingMaterializer` maybe sightly more efficient?
Which I found has some weird interfere with the optimization for singleton `List` and makes me wonder if it even worth it...